### PR TITLE
Feature/2544 metadata filtering in retrieve operations

### PR DESCRIPTION
**Pull request by kzangeli**
Fixes issue #2544

### CODE REVIEW

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 1]
```diff
@@ -2181,6 +2181,7 @@ The payload may contain the following elements (all of them optional):
    JSON object with a `type` (a JSON string) and `value` (whose type depends on the `type`
    property).
    The different available scopes are described elsewhere.
++ `metadata`: a list of metadata names to include in the response. If omitted, it means "all metadata".
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Taking into account how already is described for metadata in subscriptions:

> metadata (optional): List of metadata to be included in notification messages. By default all metadata are included except the special ones described at "Special metadata in notifications" section. The string * can be used as member of this list to mean "all the custom (user) metadata".

I'd also suggest to include the text 

> The string * can be used as member of this list to mean "all the custom (user) metadata"

also in the case of this PR.

File: src/lib/serviceRoutinesV2/postBatchQuery.cpp [hunk 2]
```diff
@@ -69,6 +69,16 @@ std::string postBatchQuery
...
 
+  //
+  // This request does not support (it ignores) the URI parameter 'metadata'.
+  // Instead the metadata filter comes inside the payload.
+  // To pass the metadata filter to the rendering function, the same mechanism as for URI param is used here.
+  // This is a little bit ugly as we 'fool' the render method to believe that the filter came via URI param.
+  // However, the implementation is cleaner and no more code is needed for this, just to set the
+  // URI param with the value of the metadata filter from the payload.
+  //
+  ciP->uriParam[URI_PARAM_METADATA] = bqP->metadataV.toString();
```
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/postBatchQuery.cpp]: Yeah, what is done here is a little ugly.
The URI param (if present) is overwritten with a stringifiction of the metadata vector from the payload and sent to the next function (the one you indicate), where the stringification is parsed into a vector again ... Ugly, however even more ugly may be to add a parameter to accept either a string-vector or a string ...   This is more or less explained in the comment right above the line in the source code

File: src/lib/apiTypesV2/Entity.cpp [hunk 3]
```diff
@@ -79,15 +80,21 @@ std::string Entity::render(ConnectionInfo* ciP, RequestType requestType, bool co
   if ((oe.details == "") && ((oe.reasonPhrase == "OK") || (oe.reasonPhrase == "")))
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Entity.cpp]: Yup

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 5786206

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 4]
```diff
@@ -1185,7 +1185,7 @@ Response:
           }
         }
 
-### Retrieve entity attributes [GET /v2/entities/{entityId}/attrs{?type,attrs,options}]
+### Retrieve entity attributes [GET /v2/entities/{entityId}/attrs{?type,attrs,options,metadata}]
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 5786206

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 5]
```diff
@@ -1127,7 +1127,7 @@ Response:
 
 ## Entity by ID [/v2/entities/{entityId}{?type,attrs,options}]
 
-### Retrieve entity [GET /v2/entities/{entityId}{?type,attrs,options}]
+### Retrieve entity [GET /v2/entities/{entityId}{?type,attrs,options,metadata}]
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 5786206

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 6]
```diff
@@ -963,7 +963,7 @@ to keep your client decoupled from implementation details.
 
 # Group Entities
 
-### List entities [GET /v2/entities{?limit,offset,options,type,id,idPattern,typePattern,q,mq,georel,geometry,coords,attrs,orderBy}]
+### List entities [GET /v2/entities{?limit,offset,options,type,id,idPattern,typePattern,q,mq,georel,geometry,coords,attrs,orderBy,metadata}]
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 5786206

File: src/lib/apiTypesV2/Entity.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/apiTypesV2/Entity.cpp]: NTC

File: src/lib/serviceRoutinesV2/postBatchQuery.cpp (hunk 2 above)
Comment (fgalan) [inline src/lib/serviceRoutinesV2/postBatchQuery.cpp]: NTC

**Comment by fgalan** [comment]
LGTM

**Pull request merged**